src/Shared/Microsoft.AspNetCore.Razor.Test.Common/IntegrationTestAttributes.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Razor.Test.Common;
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+[XunitTestCaseDiscoverer("Microsoft.AspNetCore.Razor.Test.Common." + nameof(IntegrationFactDiscoverer), "Microsoft.AspNetCore.Razor.Test.Common")]
+internal class IntegrationTestFactAttribute : FactAttribute { }
+
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+[XunitTestCaseDiscoverer("Microsoft.AspNetCore.Razor.Test.Common." + nameof(IntegrationTheoryDiscoverer), "Microsoft.AspNetCore.Razor.Test.Common")]
+internal class IntegrationTestTheoryAttribute : TheoryAttribute { }
+
+internal class IntegrationFactDiscoverer(IMessageSink diagnosticMessageSink)
+    : FactDiscoverer(diagnosticMessageSink)
+{
+    public override IEnumerable<IXunitTestCase> Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
+    {
+        return [
+            new IntegrationTestCase(designTime: true, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod),
+            new IntegrationTestCase(designTime: false, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod)
+        ];
+    }
+}
+
+internal class IntegrationTheoryDiscoverer(IMessageSink diagnosticMessageSink)
+    : TheoryDiscoverer(diagnosticMessageSink)
+{
+    protected override IEnumerable<IXunitTestCase> CreateTestCasesForDataRow(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute, object[] dataRow)
+    {
+        return [
+            new IntegrationTestCase(designTime: true, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod, dataRow),
+            new IntegrationTestCase(designTime: false, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), discoveryOptions.MethodDisplayOptionsOrDefault(), testMethod, dataRow)
+        ];
+    }
+}
+
+internal class IntegrationTestCase : XunitTestCase
+{
+    private bool designTime;
```

Comment:
```suggestion
    private readonly bool _designTime;
```

Reply:
Can't be `readonly` annoyingly due to the way XUnit (de)serializes these.